test/assignment-10/CollateralizedVault.t.sol

```diff
@@ -0,0 +1,309 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.13;
+
+import "forge-std/Test.sol";
+import {CollateralizedVault} from "src/assignment-10/CollateralizedVault.sol";
+import {Dai} from "src/assignment-5/DAI.sol";
+import {WETH9} from "src/assignment-5/WETH.sol";
+import {IERC20} from "yield-utils-v2/token/IERC20.sol";
+import {ChainlinkPriceFeedMock} from "src/assignment-5/ChainlinkPriceFeedMock.sol";
+
+abstract contract ZeroState is Test {
+
+    event Deposit(address indexed user, uint256 amount);
+    event Borrow(address indexed user, uint256 amount);
+    event Repay(address indexed user, uint256 amount);
+    event Withdraw(address indexed user, uint256 amount);
+    event Liquidate(address indexed liquidator, address indexed user, uint256 debtAmount, uint256 collateralAmount);
+
+    using stdStorage for StdStorage;
+
+    address USER = address(1);
+    address bob = address(2);
+    CollateralizedVault vault;
+    Dai dai;
+    WETH9 weth;
+    ChainlinkPriceFeedMock priceFeedMock;
+
+    function setUp() public virtual {
+        dai = new Dai(block.chainid);
+        weth = new WETH9();
+        priceFeedMock = new ChainlinkPriceFeedMock();
+        priceFeedMock.setPrice(500000000000000); // = 1/2000
+        vault = new CollateralizedVault(address(dai), address(weth), address(priceFeedMock));
+
+        setDaiBalance(address(vault), 10000 ether);
+        setWethBalance(USER, 10 ether);
+        vm.prank(USER);
+        weth.approve(address(vault), 10 ether);
+    }
+
+    function setDaiBalance(address dst, uint256 balance) public {
+        stdstore
+            .target(address(dai))
+            .sig(dai.balanceOf.selector)
+            .with_key(dst)
+            .depth(0)
+            .checked_write(balance);
+    }
+
+    function setWethBalance(address dst, uint256 balance) public {
+        stdstore
+            .target(address(weth))
+            .sig(weth.balanceOf.selector)
+            .with_key(dst)
+            .depth(0)
+            .checked_write(balance);
+    }
+}
+
+contract ZeroStateTest is ZeroState {
```

Comment:
I've been playing with state diagrams for testing files and I find them very useful. Would you mind drawing one?

Reply:
@alcueca do you mean something like:

```
ZeroState
  UndercollateralizedDebtState
```
or
```
ZeroStateTest
  - testUnauthorizedCallbackSenderIsReverted
  - testUnauthorizedCallbackInitiatorIsReverted
  UndercollateralizedDebtState
    - testLiquidate
    - testRevertsIfNoProfit
```

or something else?

Reply:
Yeah, something like that, including as well the function calls that take you from one state to another.

Reply:
https://asciiflow.com/#/

```
 ┌────────────────┐
 │                │
 │                │
 │                │
 │                │
 │                │
 └───────┬────────┘
         │
         │
   mint(bob, 100)
         │
         │
         ▼
 ┌────────────────┐
 │                │
 │                │
 │                │
 │                │
 │                │
 └───────┬────────┘
         │
         │
  transfer(ada, 50)
         │
         ▼
 ┌────────────────┐
 │                │
 │                │
 │                │
 │                │
 │                │
 └────────────────┘
```

Reply:
In this particular case there's just 2 states so it's not that interesting:

```
 ┌─────────────────────┐
 │                     │
 │                     │
 │      ZeroState      │
 │                     │
 │                     │
 └──────────┬──────────┘
            │
            ▼
  // Alice deposits 1 WETH, and borrows 66% * 2000 DAI = 1320 DAI against it
  vm.startPrank(alice);
  weth.approve(address(vault), 1 ether);
  vault.deposit(1 ether);
  vault.borrow(1320 * 1e18);
  vm.stopPrank();

  // Then price of WETH falls so vault becomes undercollateralized
  oracle.setPrice(1e18 / 1600);

            │
            │
            ▼
┌───────────────────────────────────────────┐
│                                           │
│                                           │
│         UndercollateralizedState          │
│                                           │
│                                           │
└───────────────────────────────────────────┘
```